### ISSUE TITLE
The specified module could not be found

### ISSUE THREAD

**Issue by martintan** [justadudewhohacks/opencv4nodejs#593]
Hi, I need some help... I have been struggling with this error for 11 hours. During this time I was finally able to build opencv4nodejs successfully with the manual build (disabled auto-build) with OpenCV 3.4.6, no errors. However, my server just crashes when I require opencv4nodejs. 

I apologize for not copy/pasting the output because I was debugging on a remote desktop connection. On my home PC, I was able to install it and use it perfectly with OpenCV 2.4.13, however this error happened on my client's server computer.

I feel like I have already considered everything that is possible that can make this work so now I am lost as to what is causing the error.

Error log
![image](https://user-images.githubusercontent.com/31561845/60835342-709e6900-a1f5-11e9-8234-ffe94dd3cb58.png)

Environment variables
![image](https://user-images.githubusercontent.com/31561845/60835228-2d43fa80-a1f5-11e9-94ce-1e53379d0044.png)

I have tried `npm install opencv4nodejs` with `OPENCV_DIR=C:\OpenCV\build\x64\vc15` and `OPENCV_DIR=C:\OpenCV\build`. I get the same error for both!

PATH: 
```
C:\Python27\;C:\Python27\Scripts;%SystemRoot%\system32;%SystemRoot%;%SystemRoot%\System32\Wbem;%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\;C:\Program Files\Git\cmd;C:\ProgramData\chocolatey\bin;C:\Program Files\CMake\bin;%OPENCV_DIR%;%OPENCV_BIN_DIR%;%OPENCV_LIB_DIR%;C:\Program Files\nodejs\
```

I have also tried checking the Dependency walker. No missing DLLs for OpenCV but it shows some missing DLLs that are not related to OpenCV.
![image](https://user-images.githubusercontent.com/31561845/60835784-9bd58800-a1f6-11e9-8723-63fac0aa9593.png)

*OpenCV version* (example 3.4.1): 3.4.6

*With OpenCV-contrib?* (extra modules): no

*OS*: Windows Server 2012 R2

**Comment by justadudewhohacks** [comment]
> I have also tried checking the Dependency walker. No missing DLLs for OpenCV but it shows some missing DLLs that are not related to OpenCV.

Hmm, hard to say. Did you maybe build some stuff into OpenCV, which requires external dlls?

**Comment by martintan** [comment]
I downloaded the OpenCV prebuilt executables from the website and extracted it to C:/OpenCV. I tried different versions 3.4.6/2.4.13/4 and it seems that the versions didn't matter. Earlier I even tried to just copy the OpenCV files from this folder called opencv-build that was generated by the auto build script and it seemed to work but still nothing fixes the error

**Comment by justadudewhohacks** [comment]
>  I tried different versions 3.4.6/2.4.13/4 and it seems that the versions didn't matter.

This indicates to me, that your installation in C:/OpenCV is not used at all, because you would have gotten compilation error when using OpenCV 2 to compile opencv4nodejs.

The error in "process.dlopen", which you are receiving is a runtime error (not a compilation error) and it means that there are dlls missing or invalid (the .lib files used to compile opencv4nodejs do not match the dlls in your path).

I would suggest a clean installation. Clear opencv-build from your node_modules folder and download OpenCV 3 or OpenCV 4 again. Then recompile the package and do not forget to set OPENCV4NODEJS_DISABLE_AUTOBUILD.

**Comment by martintan** [comment]
This is what happens when I try to do a clean installation:
![image](https://user-images.githubusercontent.com/31561845/60847115-25478300-a214-11e9-8170-8c416c66a881.png)
![image](https://user-images.githubusercontent.com/31561845/60847076-0f39c280-a214-11e9-888f-fe2b06e97f90.png)

**Comment by martintan** [comment]
Now I tried installing via autobuild script. I also removed the OPENCV environment variables from the system path before installing because I read an issue where it may conflict with autobuild version of OpenCV. It installs OpenCV without issues. However, still the same error "specified module could not be found". Also when I check dependency walker, now the opencv4nodejs.node has missing dependencies OPENCV_CALIB3D346.DLL, OPENCV_CORE346.DLL, etc.

**Comment by justadudewhohacks** [comment]
Can you verify that OPENCV4NODEJS_DISABLE_AUTOBUILD is set for the process that requires opencv4nodejs? If so, then the binary path to the OpenCV binaries of opencv-build should be added automatically [here](https://github.com/justadudewhohacks/opencv4nodejs/blob/master/lib/cv.js#L5-L11). Does manually adding this path to the process path variable fix the issue?

**Comment by martintan** [comment]
I tried to delete node-modules and do a clean `npm install` and deleted OPENCV4NODEJS_DISABLE_AUTOBUILD from my environment variables since the lines you mentioned in `cv.js` append the path to `process.env.path`.
```
// ensure binaries are added to path on windows
if (!opencvBuild.isAutoBuildDisabled() && process.platform === 'win32') {
  // append opencv binary path to node process
  if (!process.env.path.includes(opencvBuild.opencvBinDir)) {
    process.env.path = `${process.env.path};${opencvBuild.opencvBinDir};`
  }
}
```
I also tried to append the binary path to `process.env.path` in my code before I required opencv4nodejs. 
```
process.env.path = `${process.env.path};C:\\lbc-backend\\node_modules\\opencv-build\\opencv\\build\\bin\\Release`;
const os = require("os");
const cv = require("opencv4nodejs");
const PDFImage = require("pdf-image").PDFImage;
const Quagga = require("quagga").default;
const performance = require("perf_hooks").performance;
// const jsBarcodeReader = require("javascript-barcode-reader");
// const { BrowserBarcodeReader } = require("@zxing/library");

app.post("/locate_barcode", async (req, res) => {
  const { filePath } = req.body;
  try {
    const t0 = performance.now();
    console.log(`Converting PDF to image: ${filePath}`);
    const pdfImage = new PDFImage(filePath.replace(/\//g, "\\"), {
      outputDirectory: os.tmpdir(),
      convertOptions: {
        "-density": "300",
        "-quality": "85"
      },
      convertExtension: "jpg"
    });
```
I also put the same binary path as my OPENCV_DIR variable and put it in my system path just in case. I removed the other OPENCV environment variables because I use auto-build for this attempt.
![image](https://user-images.githubusercontent.com/31561845/60910419-532fd480-a2b3-11e9-80db-bd63fc6db76c.png)
![image](https://user-images.githubusercontent.com/31561845/60910431-5cb93c80-a2b3-11e9-9890-a5f90b5225e7.png)

Still the same error... :(
![image](https://user-images.githubusercontent.com/31561845/60910487-87a39080-a2b3-11e9-8965-fab204e72bcf.png)

**Comment by martintan** [comment]
I solved my problem as of this moment thanks to the dependency walker! It is actually a very important note that the problem computer's OS is running Windows Server 2012 R2. 

The problem was not at all with the paths not being recognized or set on the system path because I kept double-checking my environment path, even console logging the `process.env.path` to check if the path has been set for the binaries in `opencv-build`. The problem was that for some reason there are Windows DLLs that are required by OpenCV which are not installed in Windows Server.

I compared the dependency walker missing DLLs with my home PC's missing DLLs and spotted several DLLs that were missing from the problem computer's opencv4nodejs.node. These we're MF.DLL, MFPLAT.DLL, MFCORE.DLL, and multiple other DLLs that start with "MF". According to research, it is related to Microsoft Media Foundation which is not by default part of Windows Server. I installed via [Windows Server Essentials Media Pack](https://www.microsoft.com/en-ph/download/details.aspx?id=40837). After installing and restarting, it works.

Thank you for your work on this project. I would like to apologize for the inconvenience. I am also trying to learn N-API because I want to use a C library in Electron.
